src/i2c.rs

```diff
@@ -0,0 +1,375 @@
+use {
+    crate::{
+        dprintln,
+        target::{self, i2c, GPIO, I2C0, I2C1, IO_MUX},
+    },
+    core::{fmt::Write, ops::Deref},
+};
+
+unsafe fn gpio_matrix_out(gpio: u32, signal_idx: u32, out_inverted: bool, oen_inverted: bool) {
+    let base_address = 0x3FF44530; // GPIO_FUNC0_OUT_SEL_CFG_REG
+    let store_address = (base_address + 4 * gpio) as *mut u32;
+
+    let mut value = signal_idx;
+
+    if out_inverted {
+        value = value | 0b100_0000_0000;
+    }
+
+    if oen_inverted {
+        value = value | 0b1_0000_0000_0000;
+    }
+
+    *store_address = value;
```

Comment:
Here should be a volatile write too.

Reply:
Agree, however tried this already and didn't help.